torch/csrc/api/include/torch/data/samplers/distributed.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <torch/data/samplers/base.h>
+
+#include <cstddef>
+#include <vector>
+
+namespace torch {
+namespace serialize {
+class OutputArchive;
+class InputArchive;
+} // namespace serialize
+} // namespace torch
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+/// A `Sampler` that selects a subset of indices to sample from and defines a
+/// sampling behavior. In a distributed setting, this selects a subset of the
+/// indices depending on the provided num_replicas and rank parameters. The
+/// `Sampler` performs a rounding operation based on the `allow_duplicates`
+/// parameter to decide the local sample count.
+class DistributedSampler : public Sampler<> {
+ public:
+  TORCH_API DistributedSampler(
+      size_t size,
+      size_t num_replicas = 1,
+      size_t rank = 0,
+      bool allow_duplicates = true);
+
+  /// Set the epoch for the current enumeration. This can be used to alter the
+  /// sample selection and shuffling behavior.
+  TORCH_API void set_epoch(size_t epoch);
+
+  /// Return the number of samples available for this instance. This is
+  /// different to the `size` as each sampler needs to handle only a subset of
+  /// samples.
+  TORCH_API size_t local_sample_count();
+
+ protected:
+  size_t size_;
+  size_t num_replicas_;
+  size_t rank_;
+  size_t epoch_;
+  size_t local_sample_count_;
```

Comment:
It doesn't seem like `size_` has to be part of this class. `local_sample_count_` can be computed with a single division operation, so it seems very wasteful to pay 8B for caching it. Let's make base classes lean unless we have values that are expensive to compute.

Reply:
The change will do these
instead of local_sample_count_, the class need to add allow_duplicate_ as a member variable
instead of a member variable, all calculatiions will use a method call which has integer divisions. 
Accessing the count is not on the critical method (next()) so not sure what we are optimizing here.